storage/src/main/java/io/delta/storage/S3SingleDriverLogStore.java

```diff
@@ -0,0 +1,332 @@
+            Path resolvedPath) throws IOException, URISyntaxException {
+        final boolean useCache = !isInitialVersion(resolvedPath);
+        final Iterator<FileStatus> iter = listFromInternal(fs, resolvedPath, useCache);
+        if (!iter.hasNext()) return false;
+
+        return iter.next().getPath().getName().equals(resolvedPath.getName());
+    }
+
+    ////////////////////////
+    // Public API Methods //
+    ////////////////////////
+
+    @Override
+    public void write(
+            Path path,
+            Iterator<String> actions,
+            Boolean overwrite,
+            Configuration hadoopConf) throws IOException {
+        try {
+            final FileSystem fs = path.getFileSystem(hadoopConf);
+            final Path resolvedPath = stripUserInfo(fs.makeQualified(path));
+            final Path lockedPath = getPathKey(resolvedPath);
+            acquirePathLock(lockedPath);
+
+            try {
+                if (exists(fs, resolvedPath) && !overwrite) {
+                    throw new java.nio.file.FileAlreadyExistsException(
+                        resolvedPath.toUri().toString()
+                    );
+                }
+
+                final CountingOutputStream stream =
+                    new CountingOutputStream(fs.create(resolvedPath, overwrite));
+
+                while (actions.hasNext()) {
+                    stream.write((actions.next() + "\n").getBytes(StandardCharsets.UTF_8));
+                }
+                stream.close();
+
+                // When a Delta log starts afresh, all cached files in that Delta log become
+                // obsolete, so we remove them from the cache.
+                if (isInitialVersion(resolvedPath)) {
+                    final List<Path> obsoleteFiles = writtenPathCache
+                        .asMap()
+                        .keySet()
+                        .stream()
+                        .filter(p -> p.getParent() == lockedPath.getParent())
+                        .collect(Collectors.toList());
+
+                    writtenPathCache.invalidateAll(obsoleteFiles);
+                }
+
+                // Cache the information of written files to help fix the inconsistency in future
+                // listings
+                writtenPathCache.put(
+                    lockedPath,
+                    new FileMetadata(stream.getCount(), System.currentTimeMillis())
+                );
+            } catch (org.apache.hadoop.fs.FileAlreadyExistsException e) {
+                // Convert Hadoop's FileAlreadyExistsException to Java's FileAlreadyExistsException
+                throw new java.nio.file.FileAlreadyExistsException(e.getMessage());
+            } finally {
+                releasePathLock(lockedPath);
+            }
+        } catch (java.net.URISyntaxException e) {
+            // Hide this internal error and re-throw using public-api wrapper
+            throw new IOException("S3SingleDriverLogStore: java.net.URISyntaxException", e);
```

Comment:
@zsxwing would it be better to update `LogStore.java`s interfaces to throw `Exception` instead?